lib/manager/components/transform/Transform.js

```diff
@@ -0,0 +1,63 @@
+// @flow
+
+import { Component } from 'react'
+
+import type {
+  Feed,
+  FeedTransformation as FeedTransformationType
+} from '../../../types'
+
+export type TransformProps<StateType> = {
+  feedSource: Feed,
+  index: number,
+  onSave: (StateType, number) => void,
+  onValidationErrors: (Array<string>) => void,
+  transformation: FeedTransformationType
+}
+
+/**
+ * Generic Transform component that holds helper methods for derived components.
+ */
+export default class Transform<StateType> extends Component<TransformProps<StateType>, StateType> {
```

Comment:
[Inheritance is not recommended for creating React components](https://reactjs.org/docs/composition-vs-inheritance.html). In here it seems like the `updateState` is the main way components communicate with this component which then will call `onValidationErrors`. Perhaps one way to go to avoid inheritance is to create a component that handles all of the differences between each transformation type. But by then this all might start to look like the beginning of a form state management component which kind of brings up the question of why aren't we using something like Formik. Also, maybe things would be easier if all these transformRules were a separate mongo collection instead of a list within a feed source record.

Reply:
Yea, this seems like a bit of a radical change to the structure of the components that might have benefitted from a technical discussion before taking the plunge.

Reply:
I changed that in https://github.com/ibi-group/datatools-ui/pull/649/commits/54f1aac3573cc5af4575bd839afb116708e626e8, although it creates some code duplication for managing the state.